packages/@aws-cdk/cdk/lib/cloudformation/fn.ts

```diff
@@ -123,15 +124,29 @@ export class FnConcat extends FnJoin {
       } else if (isConcatIntrinsic(el)) {
         const values = concatIntrinsicValues(el);
         listOfValues.splice(i, 1, ...values);
-        i += values;
+        i += values.length;
+      } else if (i > 0
+              && typeof listOfValues[i - 1] === 'string'
+              && typeof listOfValues[i] === 'string'
+              && !unresolved(listOfValues[i - 1])
+              && !unresolved(listOfValues[i])) {
+        // Previous & current both are plain strings, concatenating them right here...
```

Comment:
Good effort, but this doesn't actually save you IAM size. By the time the policy statement gets to IAM everything will have been concatted already. So, you know. You could have saved yourself some broken tests :).

Now that it's implemented, it does save somewhat of the CFN template size, though. Not opposed to keeping it.

Reply:
In fact you should probably do this optimization round *after* the argument splatting that happens already. Right now, I believe the following join not be inlined, right?

```ts
new FnConcat('x', new FnConcat('y'));
```
But it could be.

Reply:
The reason I made this was because I found it made the expected templates _much_ harder to read... That wasn't so much for a size reduction goal here. I have re-written this part so it works for more cases of `Fn::Join`, actually.